peloton/update_project.py

```diff
@@ -0,0 +1,1371 @@
+            project_fields = ProjectFieldsQuery()
+            field_types = project_fields.data_frame[project_fields.cols.data_type]
+
+        def value_kwarg(field_id_: str, value) -> dict:
+            """
+            Use the data types in the GH schema to cast each value correctly.
+            """
+            if field_types is not None:
+                field_type_ = field_types.loc[field_id_]
+                if field_type_ is github_schema.ProjectV2FieldValue.date:
+                    # Convert a date-time string to a date string.
+                    value = value[:10]
+                return {field_type_.name: field_type_.type(value)}
+            else:
+                message = (
+                    "Problem with updating project item fields: "
+                    "cannot access information on field data types."
+                )
+                raise ValueError(message)
+
+        for _, row in data_frame.iterrows():
+            kwargs = (
+                self._inputs_default
+                |
+                dict(item_id=row[project_cols.id_in_project])
+            )
+
+            # The field in the project that contains the ID of the linked item.
+            link_field = "PVTF_lADOABU7f84ALhAIzgPKqWg"
+            link_kwargs = dict(
+                field_id=link_field,
+                value=value_kwarg(link_field, row.name),
+            )
+            inputs.append((kwargs | link_kwargs))
+
+            for field_id, col_name in issue_cols.project_field_map.items():
+                extra_kwargs = dict(field_id=field_id)
+                if col_name in row.index:
+                    if not pd.isnull(row[col_name]):
+                        # Create a `value` input argument if there is a value
+                        #  available.
+                        extra_kwargs["value"] = (
+                            value_kwarg(field_id, row[col_name])
+                        )
+                    inputs.append((kwargs | extra_kwargs))
+
+        return inputs
+
+    def _add_selector(
+        self,
+        mutation: sgqlc.operation.Operation,
+        index: int
+    ) -> None:
+        inputs = self._inputs[index]
+        if "value" in inputs:
+            op = mutation.update_project_v2_item_field_value
+        else:
+            # No value found, so clear the field instead.
+            op = mutation.clear_project_v2_item_field_value
+
+        op(input=inputs, __alias__=self._aliases[index]).client_mutation_id()
+
+
+###############################################################################
+
+
+def log_data_frame_items(
+    data_frame: pd.DataFrame,
+    message: str,
+    url_col: str = None
+) -> None:
+    """
+    Logging of HOW MANY, and WHAT, items were involved in a script step.
+    """
+    logging.info(message + f": {len(data_frame)} items")
+    if url_col is not None and not data_frame.empty:
+        logging.debug(
+            message + ":\n" + "\n".join(data_frame[url_col].tolist())
```

Comment:
```suggestion
            f"{message}:\n{'\n'.join(data_frame[url_col].tolist())}"
```

Reply:
Think I tried this one, but f-strings don't like backslashes. Here's something slightly different:

888860a